### PR TITLE
CP-25017: Ask xapi whether to use TLS or NOTLS.

### PR DESCRIPTION
**Pull request by thomassa**
* Remove the command-line "--no-tls" option.
* Instead we make an inquiry over a XenAPI session to find out about
  network purposes.

### CODE REVIEW

**Comment by thomassa** [comment]
This needs to be built against a xapi with the latest changes that introduce Network.purpose; otherwise this will fail to compile.
I've dev-tested based on using tags instead of purposes, but I'll retest properly with purposes on Monday.

File: src/main.ml
Comment (gaborigloi) [inline src/main.ml]: Let's move `Local_xapi_session` into its own `.ml` file, because it's not strictly related to cleanup any more.

File: src/main.ml
Comment (gaborigloi) [inline src/main.ml]: Here we could use `Lwt_log.error`.

File: src/main.ml
Comment (gaborigloi) [inline src/main.ml]: Here we could use `Lwt_log.warning`.

File: src/main.ml
Comment (gaborigloi) [inline src/main.ml:118]: We could try to make this a lazy value so that we don't init TLS when we don't need it. Although that would make testing a bit harder. So let's not do it for now.

**Review by gaborigloi — COMMENTED** [review]
Looks good to me. Just some minor comments above.

File: src/main.ml
Comment (thomassa) [reply src/main.ml]: Definitely better that way.

**Comment by thomassa** [comment]
I'll make those changes on Monday. Tired now.

**Comment by mseri** [comment]
Do we need to release a new xen-api-client that includes the Network.purpose change? I can easily do it on Monday morning if needed

**Comment by gaborigloi** [comment]
@mseri yes, we need to release it on the feature/CBT branch though, because we're using some new APIs here that aren't in master yet.

**Comment by mseri** [comment]
Mh, maybe we should add a CBT branch to xen-api-client with the new api, and link it in a custom xs-opam until the merge

**Comment by gaborigloi** [comment]
Sounds good! We don't even have to make a custom xs-opam, it was opam-pinned to a given version, together with `nbd`, in our `xs-opam-repo.spec`, and that worked fine so far, and it's also very easy to update the pinned version or to remove the pinning from the spec file.

**Comment by thomassa** [comment]
I've rebased again on top of master, and made changes in response to @gaborigloi's comments.
They're all separate commits for ease of review: should squash before merging. (Might keep the module-reshuffle in a commit on its own.)

**Comment by mseri** [comment]
The compilation issue can be fixed in a later pull-request. The changes in https://github.com/xapi-project/xen-api-client/pull/58 will allow to easily generate and pin API clients for different xapi branches.

**Comment by mseri** [comment]
@thomassa You can squash and merge using the github merge button, I think I would use that in this case

File: lib/local_xapi_session.ml
Comment (gaborigloi) [inline lib/local_xapi_session.ml]: We could move the docstring from here to the `.mli`.

File: lib/jbuild
Comment (gaborigloi) [inline lib/jbuild]: We could have the modules in their own line and sorted alphabetically, as for `vbd_store`.

**Review by gaborigloi — APPROVED** [review]

**Pull request merged**
